File: Algorithms/src/ywy/level1/ClawCrane.java
package ywy.level1;

public class ClawCrane {
	public int solution(final int[][] board, final int[] moves) {
		int answer = 0;
		final int[] temp = new int[moves.length];
		int index = 0;
		final int range = moves.length;
		int col = -1;
		int item = 0;
		int prev = 0;
		int next = 0;
		int prevItem = -1;
		
		do {
			col = moves[index++];
			item = boardSearch(board, --col);
			if (item > 0) {
				temp[next++] = item;
				if (next > 1) {
					prev = prev < 0 ? 0 : prev;
					prevItem = temp[prev++];
					if (item == prevItem) {
						temp[--next] = 0;
						temp[--prev] = 0;
						next = prev--;
						answer += 2;
					}
				}
			}
		} while(index < range);
		return answer;
	}
	
	private int boardSearch(final int[][] board, int col) {
		int result = -1;
		int row = 0;
		int range = board.length;
		do {
			result = board[row][col];
			if (result > 0) {
				board[row][col] = 0;
				return result;
			}
		} while((++row) < range);
		return result;
	}
}

File: Algorithms/src/ywy/javascript/level1/ClawCrane.js
const solution = (board, moves) => {
    const stack = [];
    let next = 0;
    let item = -1;
    let answer = 0;

    moves.forEach( col => {
        col--;
        item = boardSearch(board, col);
        if (item > 0) {
            if (next > 0 && item === stack[next - 1]) {
                    stack[--next] = 0;
                    answer += 2;
                    return;
            }    
            stack[next++] = item;
        }
    });
    return answer;
};

const boardSearch = (board, col) => {
    let index = 0;
    let range = board.length;
    let item = 0;
    do {
        item = board[index][col];
        if (item > 0) {
            board[index][col] = 0;
            return item;
        }
    } while((++index) < range);
    return item;
};